controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -259,91 +259,133 @@ func MatchesKubernetesVersion(kubernetesVersion string) Func {
+	kcpConfig := getAdjustedKcpConfig(kcp, machineConfig)
 
-		// Machine's JoinConfiguration Discovery is set to an empty object by KubeadmConfig controller if KCP is Discovery is nil, also
-		// adopted machines may have an unrelated discovery setting so ignore discovery completely when comparing JoinConfigurations.
-		emptyDiscovery := kubeadmv1.Discovery{}
-		if kcpConfigSpecLocal.JoinConfiguration != nil {
-			kcpConfigSpecLocal.JoinConfiguration.Discovery = emptyDiscovery
-		}
-		if bootstrapObj.Spec.JoinConfiguration != nil {
-			bootstrapObj.Spec.JoinConfiguration.Discovery = emptyDiscovery
-		}
+	// cleanups all the fields that are not relevant for the comparison.
+	cleanupConfigFields(kcpConfig, machineConfig)
 
-		// Machine's JoinConfiguration ControlPlane is set to an empty object by KubeadmConfig controller if KCP is ControlPlane is nil.
-		// Set Machine's ControlPlane to nil to avoid
-		if kcpConfigSpecLocal.JoinConfiguration != nil && kcpConfigSpecLocal.JoinConfiguration.ControlPlane == nil {
-			bootstrapObj.Spec.JoinConfiguration.ControlPlane = nil
-		}
+	return reflect.DeepEqual(&machineConfig.Spec, kcpConfig)
+}
 
-		// If KCP's join NodeRegistration is empty, set machine's node registration to empty as no changes should trigger rollout.
-		emptyNodeRegistration := kubeadmv1.NodeRegistrationOptions{}
-		if kcpConfigSpecLocal.JoinConfiguration != nil && reflect.DeepEqual(kcpConfigSpecLocal.JoinConfiguration.NodeRegistration, emptyNodeRegistration) {
-			bootstrapObj.Spec.JoinConfiguration.NodeRegistration = emptyNodeRegistration
-		}
+// getAdjustedKcpConfig takes the KubeadmConfigSpec from KCP and applies the transformations required
+// to allow a comparison with the KubeadmConfig referenced from the machine.
+// NOTE: The KCP controller applies a set of transformations when creating a KubeadmConfig referenced from the machine,
+// mostly depending on the fact that the machine was the initial control plane node or a joining control plane node.
+// In this function we don't have such information, so we are making the KubeadmConfigSpec similar to the KubeadmConfig.
+func getAdjustedKcpConfig(kcp *controlplanev1.KubeadmControlPlane, machineConfig *bootstrapv1.KubeadmConfig) *bootstrapv1.KubeadmConfigSpec {
+	kcpConfig := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+
+	// Machine's join configuration is nil when it is the first machine in the control plane.
+	if machineConfig.Spec.JoinConfiguration == nil {
+		kcpConfig.JoinConfiguration = nil
+	}
+
+	// Machine's init configuration is nil when the control plane is already initialized.
+	if machineConfig.Spec.InitConfiguration == nil {
+		kcpConfig.InitConfiguration = nil
+	}
+
+	return kcpConfig
+}
 
-		return reflect.DeepEqual(bootstrapObj.Spec, *kcpConfigSpecLocal)
+// cleanupConfigFields cleanups all the fields that are not relevant for the comparison.
+func cleanupConfigFields(kcpConfig *bootstrapv1.KubeadmConfigSpec, machineConfig *bootstrapv1.KubeadmConfig) {
```

Comment:
Can we merge this with the above function? Seems they're called one after another anyway

Reply:
IMO the current code structure gives a better separation of concerns and simpler testability.
However, no problem to change if you prefer this way

Reply:
Either that, or call `cleanupConfigFields` within the other function. It seems a little overkill to have a single function that's being called in only one place. Inlining keeps the context clear.